compiler/src/dotty/tools/dotc/transform/Erasure.scala

```diff
@@ -675,4 +679,485 @@ object Erasure {
+      } else {
+        // implement new spec for erasure of refined types.
+        def isUnshadowed(psym: Symbol) =
+          !(psyms exists (qsym => (psym ne qsym) && (qsym isSubClass psym)))
+        val cs = parents.iterator.filter { p => // isUnshadowed is a bit expensive, so try classes first
+          val psym = p.typeSymbol
+          psym.ensureCompleted()
+          psym.isClass && !psym.is(Flags.Trait) && isUnshadowed(psym)
+        }
+        (if (cs.hasNext) cs else parents.iterator.filter(p => isUnshadowed(p.typeSymbol))).next()
+      }
+    }
+  }
+
+  /* Drop redundant types (ones which are implemented by some other parent) from the immediate parents.
+   * This is important on Android because there is otherwise an interface explosion.
+   */
+  private def minimizeParents(cls: Symbol, parents: List[Type])(implicit ctx: Context): List[Type] = if (parents.isEmpty) parents else {
+    // val requiredDirect: Symbol => Boolean = requiredDirectInterfaces.getOrElse(cls, Set.empty)
+    var rest   = parents.tail
+    var leaves = collection.mutable.ListBuffer.empty[Type] += parents.head
+    while (rest.nonEmpty) {
+      val candidate = rest.head
+      val candidateSym = candidate.typeSymbol
+      // val required = requiredDirect(candidateSym) || !leaves.exists(t => t.typeSymbol isSubClass candidateSym)
+      val required = !leaves.exists(t => t.typeSymbol.isSubClass(candidateSym))
+      if (required) {
+        leaves = leaves filter { t =>
+          val ts = t.typeSymbol
+          !(ts.is(Flags.Trait) || ts.is(Flags.PureInterface)) || !candidateSym.isSubClass(ts)
+          // requiredDirect(ts) || !ts.isTraitOrInterface || !candidateSym.isSubClass(ts)
+        }
+        leaves += candidate
+      }
+      rest = rest.tail
+    }
+    leaves.toList
+  }
+
+  private def hiBounds(bounds: TypeBounds)(implicit ctx: Context): List[Type] = bounds.hi.widenDealias match {
+    case AndType(tp1, tp2) => tp1 :: tp2 :: Nil
+    case tp => tp :: Nil
+  }
+
+  /** Arrays despite their finality may turn up as refined type parents,
+    *  e.g. with "tagged types" like Array[Int] with T.
+    */
+  private def unboundedGenericArrayLevel(tp: Type)(implicit ctx: Context): Int = tp match {
+    case GenericArray(level, core) if !(core <:< defn.AnyRefType) =>
+      level
+    case AndType(tp1, tp2) =>
+      Math.max(unboundedGenericArrayLevel(tp1), unboundedGenericArrayLevel(tp2))
+    case _ =>
+      0
+  }
+
+  // only refer to type params that will actually make it into the sig, this excludes:
+  // * higher-order type parameters
+  // * type parameters appearing in method parameters
+  // * type members not visible in an enclosing template
+  private def isTypeParameterInSig(sym: Symbol, initialSymbol: Symbol)(implicit ctx: Context) = {
+    !sym.isHigherOrderTypeParameter &&
+      sym.isTypeParameterOrSkolem && (
+      (initialSymbol.enclClassChain.exists(sym isContainedIn _)) ||
```

Comment:
I believe equivalent, and faster would be:

    sym.isContainedIn(initialSymbol.topLevelClass)

Reply:
That also means we don't need `enclClassChain`.